client/my-sites/plans-grid/components/comparison-grid/index.tsx

```diff
@@ -892,6 +1017,32 @@ const ComparisonGrid = ( {
 		setVisiblePlans( newVisiblePlans );
 	}, [ isLargeBreakpoint, isMediumBreakpoint, isSmallBreakpoint, displayedGridPlans, isInSignup ] );
 
+	const visibleGridPlans = useMemo(
+		() =>
+			visiblePlans.reduce( ( acc, planSlug ) => {
+				const gridPlan = displayedGridPlans.find( ( gridPlan ) => gridPlan.planSlug === planSlug );
+
+				if ( gridPlan ) {
+					acc.push( gridPlan );
+				}
+
+				return acc;
+			}, [] as GridPlan[] ),
+		[ visiblePlans, displayedGridPlans ]
+	);
+
+	const onPlanChange = useCallback(
```

Comment:
I think you are making a good observation. This was memoized from before btw, just moved here to bring related code a little closer. 

I do notice several effective memoized calls though from everything else that's hoisted at the upper level e.g. interacting with tooltips or scrolling will recreate this. So there are potential gains, then there are loses from interacting with `visiblePlans`. I'm not sure - I'd keep it cause I doubt we'd be inclined to re-memo if/when some new interaction is added.

I'd be more happy to remove though if there was a noticeable gain? 🤔

Reply:
> This was memoized from before btw, just moved here to bring related code a little closer.

Got it! Thanks for making these improvements!

We can eliminate the dependency array and also the need for memoization by rewriting the function like this:
```js
const onPlanChange = ( currentPlan: PlanSlug, event: ChangeEvent< HTMLSelectElement > ) => {
  const newPlan = event.currentTarget.value;
  
  setVisiblePlans( visiblePlans => visiblePlans.map( ( plan ) =>
	  plan === currentPlan ? ( newPlan as PlanSlug ) : plan
  ) );
};
```

So there will definitely be a net loss due to the unnecessary `useCallback`, but I'm not sure if it's noticeable or not!

Reply:
I am a little confused. Maybe you can guide me a little as I'm probably missing something, seeing how confident you are. How is this suggestion _not_ going to recreate this reference on every render (and force rerender anything that consumes it)?

Reply:
> How is this suggestion not going to recreate this reference on every render (and force rerender anything that consumes it)?

You're right, this will re-create the reference on every render. However, in most cases (excluding the initial re-renders), the re-rendering of the child components that consume this callback will most likely be caused due to a change in the parent component's state - `visiblePlans`, `activeTooltipId`, etc. We are not wrapping the child components with `memo`, so they'll be re-rendered anyway and preserving the reference of the function using `useCallback` might not have any impact. But react will have to invoke `useCallback` itself on every render, along with applying its diffing logic for the dependency array, so that's an additional cost as well.

If we separate out the `useCallback`, we can rewrite it like this, which is equivalent to the original one.

```js
const onPlanChange = ( currentPlan: PlanSlug, event: ChangeEvent< HTMLSelectElement > ) => {
  const newPlan = event.currentTarget.value;
  
  setVisiblePlans( visiblePlans => visiblePlans.map( ( plan ) =>
	  plan === currentPlan ? ( newPlan as PlanSlug ) : plan
  ) );
};
const onPlanChangeCallback = useCallback( onPlanChange, [] );
```

If `useCallback` does not have any downstream impact on reducing child re-renders, then there is a cost attached to this. (although very small! 🙂 )

I could have also missed something here, so happy to learn more about this!

Reply:
> We are not wrapping the child components with memo, so they'll be re-rendered anyway and preserving the reference of the function using useCallback might not have any impact.

I can't do everything in this PR. We can bring individualized component memoization down the line. I think the important thing, right now, is to ensure that references to data and callbacks are stable. I would rather we are consistent on anything higher up the tree, otherwise, it's all a bit hit-and-miss situation (little bits here and little bits there). I totally understand the costs, just don't think they are significant to block further optimizing things, right now.

```
const onPlanChange = ( currentPlan: PlanSlug, event: ChangeEvent< HTMLSelectElement > ) => {
  const newPlan = event.currentTarget.value;
  
  setVisiblePlans( visiblePlans => visiblePlans.map( ( plan ) =>
	  plan === currentPlan ? ( newPlan as PlanSlug ) : plan
  ) );
};
const onPlanChangeCallback = useCallback( onPlanChange, [] );
```

This is ok, as long as no other dependencies are introduced into this handler. The only other way I see to avoid `useCallback` call is to extract handler out of component scope and pass everything through as props. So basically to remove the closure entirely. Different maintenance and design concerns there.

Reply:
I'm not an expert, certainly not at the point of devilizing a case of `useCallback` on a handler passed down the tree from main component. I can totally see these handlers evolving too 🤷‍♂️